File: golang/src/Acme/Application/Command/Handler/CreateUserHandler.go
package Handler

import (
	. "../../../../SharedKernel/Application/Command"
	. "../../../Domain"
	. "../../../Domain/Model"
	. "../../Command"
	"errors"
)

type CreateUserHandler struct {
	users Users
}

func NewCreateUserHandler(users Users) *CreateUserHandler {
	return &CreateUserHandler{users: users}
}

func (handler *CreateUserHandler) Handle(command Command) error {
	createUserCommand := command.(*CreateUserCommand)
	if handler.users.Exists("john@doe.com") {
		return errors.New("current user exists")
	}

	user := NewUser(
		createUserCommand.Email(),
		createUserCommand.Password(),
		createUserCommand.FirstName(),
		createUserCommand.LastName())
	handler.users.Add(user)

	return nil
}

File: typescript/src/Acme/Application/Command/Handler/CreateUserHandler.ts
import {Handler} from "../../../../SharedKernel/Application/Command/Handler/Handler";
import {Command} from "../../../../SharedKernel/Application/Command/Command";
import {CreateUserCommand} from "../CreateUserCommand";
import {User} from "../../../Domain/Model/User";
import {Users} from "../../../Domain/Users";

export class CreateUserHandler implements Handler {
    private users: Users;

    constructor(users: Users) {
        this.users = users;
    }

    handle(command: Command): void {
        if (!(command instanceof CreateUserCommand)) {
            throw new Error('Invalid command type');
        }

        if (this.users.exists(command.email)) {
            throw new Error('Current user exists');
        }

        const user = new User(command.email, command.password, command.firstName, command.lastName);

        this.users.add(user);
    }

}